sklearn/tree/_tree.pyx

```diff
@@ -446,7 +454,8 @@ cdef class BestFirstTreeBuilder(TreeBuilder):
 
         if not is_leaf:
             splitter.node_split(impurity, &split, &n_constant_features)
-            is_leaf = is_leaf or (split.pos >= end)
+            is_leaf = (is_leaf or split.pos >= end or
+                       split.improvement + EPSILON < min_impurity_decrease)
```

Comment:
So I did this to avoid not splitting if `split.improvement` is not as low as `min_impurity_decrease` within the precision of the machine. For instance if you give `min_impurity_decrease` as `1e-7`, it does not build the tree completely as sometimes the improvement is almost equal to `1e-7`...

And I added it to the right and not left as it would give splitting the benefit of doubt (as opposed to not splitting)...